CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs

```diff
@@ -0,0 +1,437 @@
+            CampaignEvents.BeforeMissionOpenedEvent.AddNonSerializedListener(this, OnMissionStarted);
+            
+            //Parties created and destroyed
+            CampaignEvents.MobilePartyCreated.AddNonSerializedListener(this,RegisterParty);
+            CampaignEvents.MobilePartyDestroyed.AddNonSerializedListener(this,DeregisterParty);
+            
+            //Units added
+            CampaignEvents.OnPartySizeChangedEvent.AddNonSerializedListener(this,OnPartySizeChanged);
+            CampaignEvents.NewCompanionAdded.AddNonSerializedListener(this,OnCompanionAdded);
+            CampaignEvents.OnTroopRecruitedEvent.AddNonSerializedListener(this,OnTroopRecruited);
+            CampaignEvents.OnUnitRecruitedEvent.AddNonSerializedListener(this, OnPlayerRecruitedTroop);
+                //missing : OnTroopRemoved
+        }
+
+        private void OnMissionEnded(MapEvent mapEvent)
+        {
+            if (playerIsInBattle)
+            {
+                playerIsInBattle = false;
+            }
+        }
+
+        
+
+        private void OnPlayerRecruitedTroop(CharacterObject troop, int amount)
+        {
+       //     TOWCommon.Say("Player Troop Recruited");
+            _currentTroop = troop;
+            
+            PartyAttribute PlayerParty = GetAttribute(Campaign.Current.MainParty.Party.Id); 
+            AddRegularTroopToPartyAttributes(PlayerParty,troop);
+        }
+        private void OnTroopRecruited(Hero hero, Settlement settlement, Hero RecruitmentSource, CharacterObject Troop, int amount)
+        {
+            _currentTroop = Troop;
+          //  TOWCommon.Say("Troop Recruited");
+            if(hero==null) return;
+            PartyAttribute partyAttribute = GetAttribute(hero.PartyBelongedTo.Party.Id);
+            AddRegularTroopToPartyAttributes(partyAttribute, Troop);
+        }
+
+        private void OnCompanionAdded(Hero hero)
+        { 
+         //   TOWCommon.Say("Hero created");
+            _currentAddedHero = hero;
+        }
+        
+        private void AddRegularTroopToPartyAttributes(PartyAttribute party, CharacterObject troop)
+        {
+            foreach (var attribute in party.RegularTroopAttributes)
+            {
+                if (attribute.id == troop.ToString());
+                return;
+            }
+                
+            StaticAttribute staticAttribute = new StaticAttribute();
+            if (troop.Occupation == Occupation.Bandit)
+            {
+                
+                //random data   work around for rogue members
+                staticAttribute.race = "Human";
+                staticAttribute.status = "not so ready";
+                staticAttribute.MagicEffects = new List<string>();
+                staticAttribute.id = troop.ToString();
+                staticAttribute.number = party.PartyBase.MemberRoster.GetTroopCount(troop);
+                party.RegularTroopAttributes.Add(staticAttribute);
```

Comment:
All the logic here gets repeated right below it. This is unnecessary. Lines 116 to 135 could instead just be

```
StaticAttribute staticAttribute = new StaticAttribute();

staticAttribute.race = "Human";
staticAttribute.status = troop.Occupation == Occupation.Bandit ? "not so ready" : "battle ready";
staticAttribute.MagicEffects = new List<string>();
staticAttribute.id = troop.ToString();
staticAttribute.number = party.PartyBase.MemberRoster.GetTroopCount(troop);
party.RegularTroopAttributes.Add(staticAttribute);

_currentTroop = null;
```

Also, while looking at that, it appears to me that `staticAttribute.number` is an unclearly named field. What is this for? It doesn't seem like it's used anywhere in the code either other than when it's assigned. Can this just be removed from StaticAttribute?

What does the StaticAttribute.status do? Why are Bandits "not so ready" and everybody else is "battle ready"?

Reply:
number, race status was removed. For now we are just focusing on Attributes like beeing undead , human, and abilities

Reply:
the attributes got significantly reduced, and only contain a minimum of information.